node/src/components/small_network/task_manager.rs

```diff
@@ -0,0 +1,231 @@
+//!         tokio::select! {
+//!             _ = shutdown.wait_for_shutdown() => {
+//!                 // Shutdown received. Simply exit the future.
+//!                 return;
+//!             }
+//!
+//!             _ = tokio::time::sleep(Duration::from_secs(5)) => {
+//!                 println!("5 seconds have passed in task {}", n);
+//!             }
+//!         }
+//!     });
+//! }
+//!
+//! // After we have spawned the tasks, we tell them to shut down and wait for termination.
+//! let tasks_leftover = rt.block_on(
+//!     task_manager.shutdown_and_wait(Duration::from_secs(10))
+//! ).expect("shutdown failed");
+//!
+//! assert_eq!(tasks_leftover, 0);
+//! ```
+//!
+//! # `Drop`
+//!
+//! Dropping the `TaskManager` will cause all tasks to be sent a shutdown notice, but the `Drop`
+//! implementation will **not** wait them to finish.
+
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+    time::Duration,
+};
+
+use datasize::DataSize;
+use futures::{stream::FuturesUnordered, Future, StreamExt};
+use tokio::{
+    sync::watch,
+    task::{JoinError, JoinHandle},
+    time,
+};
+use tracing::{error, trace, warn};
+
+/// A shutdown receiver.
+///
+/// Can be asked for a shutdown signal via `wait_for_shutdown`.
+#[derive(DataSize, Debug, Clone)]
+pub struct ShutdownReceiver(#[data_size(skip)] watch::Receiver<()>);
+
+impl ShutdownReceiver {
+    /// Waits for the shutdown signal.
+    ///
+    /// The returned future is safe to cancel.
+    pub async fn wait_for_shutdown(&mut self) {
+        match self.0.changed().await {
+            Ok(()) => {
+                // We are using the dropping of the channel as a signal, not the channel itself!
+                error!("received value on shutdown channel, this should never happen")
+            }
+            Err(_recv_err) => {
+                // All good.
+            }
+        }
+    }
+}
+
+/// The outcome of awaiting a task for shutdown.
+#[derive(Debug)]
+enum ShutdownOutcome {
+    /// The task was shut down completely.
+    Ok,
+    /// Could not join the task, it either panicked or was explicitly cancelled.
+    JoinFailure(JoinError),
+    /// Join timeout.
+    ///
+    /// The task is still running in the background, the join handle has been lost.
+    TimedOut,
+}
+
+/// Manager for tasks.
+///
+/// Can be used to signal a group of tasks to shutdown and wait for their completion, optionally
+/// cancelling execution. All operations are "best effort", won't panic even on poisoned locks.
+///
+/// Dropping a task manager will instruct all tasks to shutdown, but the `drop` will not await their
+/// termination.
+#[derive(DataSize, Debug)]
+pub struct TaskManager {
+    /// A counter for joined tasks.
+    ///
+    /// Incremented to provide unique IDs for tasks.
+    task_counter: usize,
+    /// Join handles for all spawned tasks, mapped from task ID to join handle.
+    join_handles: Arc<Mutex<HashMap<usize, JoinHandle<()>>>>,
```

Comment:
I think we should avoid a mutex if we can.

Reply:
If we are going to use a `Mutex`, it's more idiomatic to use a `futures::Lock::Mutex`.

Reply:
> I think we should avoid a mutex if we can.

Agreed, although the only alternative I can see if not bounding the number of "dead" join handles, as we have no chance of cleaning them up if a task finishes. That seems a worse alternative.

I'm very much open to a better suggestion :)

> `futures::Lock::Mutex`

AFAIK there's no need to use a futures-aware mutex here, at least it's not recommended by `tokio::sync::Mutex` in this case at least: https://tokio-rs.github.io/tokio/doc/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use

Reply:
The only alternative I can think of is having another task with a channel that manages these things.